[19.05] Fix metadata output collection on Pulsar with new metadata strategy.

jmchilton opened the pull request:
Output fix corresponding to input fix in https://github.com/galaxyproject/galaxy/pull/7694/files. I think I had been testing with metadata retry on and didn't realize the outputs weren't returning to the correct directory.

jmchilton committed:
Fix metadata output collection on Pulsar.

bgruening commented:
I get the following in my rather complex local setup.

```
Traceback (most recent call last):
  File "lib/galaxy/jobs/runners/__init__.py", line 130, in run_next
    method(arg)
  File "lib/galaxy/jobs/runners/pulsar.py", line 517, in finish_job
    self._handle_metadata_externally(job_wrapper, resolve_requirements=True)
  File "lib/galaxy/jobs/runners/__init__.py", line 346, in _handle_metadata_externally
    kwds={'overwrite' : False})
  File "lib/galaxy/jobs/__init__.py", line 1964, in setup_external_metadata
    **kwds)
  File "lib/galaxy/metadata/__init__.py", line 107, in setup_external_metadata
    os.mkdir(metadata_dir)
OSError: [Errno 17] File exists: '/home/bag/projects/code/galaxy/database/jobs_directory/000/95/metadata'
```

jmchilton committed:
Fix portable remote metadata output files...

... three features not working together super well.

gmauro commented:
With this patch and cgroups enabled both sides (glx and pulsar), I got this error:
```
galaxy.web.framework.decorators ERROR 2019-05-31 14:40:36,297 [p:21824,w:1,m:0] [uWSGIWorker1Core1] Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/srv/galaxy/server/lib/galaxy/web/framework/decorators.py", line 282, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/srv/galaxy/server/lib/galaxy/webapps/galaxy/api/job_files.py", line 103, in create
    shutil.move(input_file.name, path)
  File "/usr/lib/python2.7/shutil.py", line 316, in move
    copy2(src, real_dst)
  File "/usr/lib/python2.7/shutil.py", line 144, in copy2
    copyfile(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 97, in copyfile
    with open(dst, 'wb') as fdst:
IOError: [Errno 2] No such file or directory: u'/srv/galaxy/jobs/000/334/metadata/__instrument_cgroup__metrics'
```

jmchilton committed:
Fix finding datatype's registry for metadata collection on Pulsar.

jmchilton commented:
Yes - so this fixes metadata for the default metadata_strategy but it does break metrics. I think @natefoo and I know this would prefer to make the fix anyway to get main working. Is this a regression on your end though? Does the job work fine and set metadata without this patch?

gmauro commented:
Yes, without this patch the job works fine.

bgruening commented:
> Does the job work fine and set metadata without this patch?

@gmauro we did not get metadata setting to work on pulsar yet, right?

jmchilton commented:
Oh I see, this would break jobs with metrics and without remote metadata... crap.

bgruening commented:
We are trying to get remote metadata working, so don't bother with this just because of us.

jmchilton commented:
Are you experiencing scaling problems that make remote metadata appealing?

I think the fact this patch breaks the default Pulsar configuration (no remote metadata) is a real problem.

bgruening commented:
> Are you experiencing scaling problems that make remote metadata appealing?

No, not yet. But we thought this is the recommended way.

> I think the fact this patch breaks the default Pulsar configuration (no remote metadata) is a real problem.

Yes, it would mean changing the default as well, which I think is ok, if it is the recommended way for production Galaxy instances.

jmchilton commented:
> No, not yet. But we thought this is the recommended way.

It isn't documented anywhere and I don't think it is worth the complexity in terms of configuration. It is a last resort IMO until we get it right in 19.09. I'd rather see Pulsar running at scale and have actual profile data on this being a problem - and then decide if just fixing the problematic datatypes are better or switching the server to remote metadata.

bgruening commented:
Ok, we can work around this on our side. Whatever you do, we can patch it on our side to make it work. Thanks John.

jmchilton committed:
Spelling fix.

jmchilton committed:
Fix job metrics collection from them returning into metadata/

Ideally all job metrics should be placed here to mimic Pulsar but this is a bug fix to an almost stable branch.

jmchilton commented:
Latest stuff should fix job metrics regression from this branch?

natefoo commented:
Appears to work for me!

jmchilton committed:
Fixes(?) to galaxy.metadata.
